components/brave_federated_learning/brave_operational_profiling.cc

```diff
@@ -0,0 +1,201 @@
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/pref_service.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "url/gurl.h"
+
+namespace brave {
+
+namespace {
+
+static constexpr char federatedLearningUrl[] = "https://fl.bravesoftware.com/";
+
+constexpr char kLastCheckedSlotPrefName[] = "brave.federated.last_checked_slot";
+constexpr char kEphemeralIDPrefName[] = "brave.federated.ephemeral_id";
+constexpr char kEphemeralIDExpirationPrefName[] =
+    "brave.federated.ephemeral_id_expiration";
+
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("brave_operational_profiling", R"(
+        semantics {
+          sender: "Operational Profiling Service"
+          description:
+            "Report of anonymized usage statistics. For more info see "
+            "TODO: https://wikilink_here"
+          trigger:
+            "Reports are automatically generated on startup and at intervals "
+            "while Brave is running."
+          data:
+            "Anonymized and encrypted usage data."
+          destination: WEBSITE
+        }
+        policy {
+          cookies_allowed: NO
+          setting:
+            "This service is enabled only when P3A is enabled and the user"
+            "has opted-in to ads."
+          policy_exception_justification:
+            "Not implemented."
+        }
+    )");
+}
+
+}  // anonymous namespace
+
+BraveOperationalProfiling::BraveOperationalProfiling(
+    PrefService* pref_service,
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory)
+    : pref_service_(pref_service), url_loader_factory_(url_loader_factory) {}
+
+BraveOperationalProfiling::~BraveOperationalProfiling() {}
+
+void BraveOperationalProfiling::RegisterLocalStatePrefs(
+    PrefRegistrySimple* registry) {
+  registry->RegisterIntegerPref(kLastCheckedSlotPrefName, -1);
+  registry->RegisterStringPref(kEphemeralIDPrefName, {});
+  registry->RegisterTimePref(kEphemeralIDExpirationPrefName, base::Time());
+}
+
+void BraveOperationalProfiling::Start() {
+  DCHECK(!simulate_local_training_step_timer_);
+  DCHECK(!collection_slot_periodic_timer_);
+
+  LoadPrefs();
+
+  simulate_local_training_step_timer_ =
+      std::make_unique<base::RetainingOneShotTimer>();
+  simulate_local_training_step_timer_->Start(
+      FROM_HERE, base::TimeDelta::FromSeconds(fake_update_duration_ * 60), this,
+      &BraveOperationalProfiling::OnSimulateLocalTrainingStepTimerFired);
+
+  collection_slot_periodic_timer_ = std::make_unique<base::RepeatingTimer>();
+  collection_slot_periodic_timer_->Start(
+      FROM_HERE, base::TimeDelta::FromSeconds(collection_slot_size_ * 60 / 2),
+      this, &BraveOperationalProfiling::OnCollectionSlotStartTimerFired);
+}
+
+void BraveOperationalProfiling::LoadParams() {
```

Comment:
thinking about it, I would get features and URL in place and not add as member variables, @iefremov wdyt?

Reply:
yeah, me too

Reply:
👍